### PR TITLE
Taxonomies: Refactor the FormTermDialog Component to allow editing

### LINKED ISSUE #8584: Update / Extract AddTerm >
Look at any updates needed to `term-tree-selector/add-term.jsx` to make it usable for the Taxonomy Manager

### PR DESCRIPTION
**Pull request by youknowriad**
* If we provide a `term` prop, the dialog shows an edit form
 * Editing a term, drops and creates a new term (API behavior)
 * Update posts after term update

**Technical aspects**

When we trigger the update API call, the API returns a new term (new ID and new Slug), attaches this term with the current site and drops the old term from the current site. The API also updates the post <-> terms relationship. 

So these changes need to be reflected on the Redux state as well. I tried to keep it simple for now and I handled those directly on the Action Creator (side effects goes there right :) ). I'm not sure if this kind of behavior (updating an entity triggers replacement of the entity and updating other related entities) is common or if it's only related to terms. If this is common on the API, we may have to think of a more advanced mechanism (similar to QueryManagers ?) to reflect changes between different entities.

closes #8584

**Testing instructions**

For now the edit form is not used anywhere on Calypso (next PR), so just make sure adding categories to posts from the Editor, works as expected.

cc @aduth @timmyc

### CODE REVIEW

File: client/blocks/term-form-dialog/index.jsx [hunk 1]
```diff
@@ -212,16 +249,19 @@ class TermFormDialog extends Component {
 				autoFocus={ false }
 				isVisible={ showDialog }
 				buttons={ buttons }
-				onClose={ this.boundCloseDialog }
-				additionalClassNames="term-form-dialog">
-				<FormSectionHeading>{ labels.add_new_item }</FormSectionHeading>
+				onClose={ this.closeDialog }
+				additionalClassNames={ 'term-form-dialog' }>
```
Comment (timmyc) [inline client/blocks/term-form-dialog/index.jsx]: Curious, why did you change this from a string to `{ '' }` here?

File: client/blocks/term-form-dialog/index.jsx [hunk 2]
```diff
@@ -141,68 +196,50 @@ class TermFormDialog extends Component {
...
+					hideTermAndChildren={ hideTermAndChildren }
 				/>
 			</FormFieldset>
 		);
 	}
 
 	render() {
 		const { isHierarchical, labels, translate, showDescriptionInput, showDialog } = this.props;
+		const { name, description } = this.state;
+		const isNew = ! this.props.term;
```
Comment (timmyc) [inline client/blocks/term-form-dialog/index.jsx]: minor but should we just assign `terms` in line 231 too?

File: client/blocks/term-form-dialog/index.jsx (hunk 1 above)
Comment (youknowriad) [inline client/blocks/term-form-dialog/index.jsx]: I'm not sure why :P I guess I changed it because I thought of adding a new computed classname and changed my mind :)
